### ISSUE TITLE
Weighted scoring for detection strategies

### ISSUE THREAD

**Issue by bkeepers** [github/linguist#2195]
This is a more detailed summary of the [proposed GSoC 2015 project](https://github.com/github/gsoc#weighted-scoring-for-detection-strategies).

Linguist currently uses [multiple strategies](https://github.com/github/linguist/blob/1a3680bba6f4676087b1a883a62cb16d29408bfb/lib/linguist/language.rb#L90-L96) to detect the language of a file, including matching on file name, heuristics, and bayesian classification. Each strategy is relatively independent, making it difficult to aggregate them into a more probable match. We would like to explore an algorithm that combines all strategies into a confidence score.

As detailed in https://github.com/github/linguist/issues/1571, it is currently impossible for linguist to return "unknown" if the file extension matches an existing language definition. For example, `.for` is defined as an extension for `Fortran` and `Forth`, but as https://github.com/github/linguist/pull/2123 explains, some `.for` files are neither of those. (/cc https://github.com/github/linguist/issues/1740, https://github.com/github/linguist/issues/2169, https://github.com/github/linguist/pull/2123, etc)

This work will likely involve changes to all of the [detection strategies](https://github.com/github/linguist/blob/1a3680bba6f4676087b1a883a62cb16d29408bfb/lib/linguist/language.rb#L90-L96), especially the [classifier](https://github.com/github/linguist/blob/master/lib/linguist/classifier.rb), to increase the accuracy in language detection and the ability to return "unknown" if the confidence score is not above a threshold.

[Detecting Programming Language from Source Code Using Bayesian Learning Techniques](http://link.springer.com/chapter/10.1007%2F978-3-319-08979-9_39) may be worth checking out.

/cc @arfon

**Comment by hTrap** [comment]
I went through all the above links provided, there was a lot of ruby in it. but the comments helped a lot. I couldn't access the paper http://link.springer.com/chapter/10.1007%2F978-3-319-08979-9_39 , is there any other resource for this paper.
Also i doubt that i am currently capable of contributing to this but i am a fast learner so if you could provide me few more resources to learning more about it and ways i can get to speed. I am finding this project really interesting and want to do justice to the project. Also i am a beginner in ruby hence i am going to read more about its oop concepts. 
Also if you could provide me small tasks to complete that would help me grasp a few basic concepts, that would really help.

**Comment by larsbrinkhoff** [comment]
Participating.

**Comment by hTrap** [comment]
I just received the mail and cannot view it on github, is it for me to
reply on?

**Comment by lazywei** [comment]
It seems to me that we can firstly deal with making a fine tune threshold for classifier -- if the score of predicted label (or language) is lower than the threshold, then we should output **unknown**.
The problem is: how can we train such threshold?

1. Use heuristic to set a fixed threshold, say 0.1 or something.
2. Use the features (i.e., the tokens) and the score as new features, and then train a binary classifier to decide whether we should accept it or not.

How do you think?

By the way, is there any chance we can implement some other more sophisticated classifier? For example, we may use [liblinear](http://www.csie.ntu.edu.tw/~cjlin/liblinear/)'s Linear SVMs or Logistic Regression. Usually, such linear classifiers perform well in such high dimensional data features, and it's intuitive to add the **unknown** as a possible prediction to such classifier -- we just add **unknown** as an option when training (I think it is not that obviously to add **unknown** as option in NB, right?). Of course, using such sophisticated classifiers will introduce more dependencies, which might be a downside.

Thanks.

**Comment by arfon** [comment]
> By the way, is there any chance we can implement some other more sophisticated classifier? F

Yes! This is definitely very interesting to us too. By the way, if this is an approach that we end up taking then it could be interesting to use Vim and Emacs modelines to gather large training sets from existing files on GitHub, for example: [this is a search](https://github.com/search?utf8=%E2%9C%93&q=%22vim%3A+set+ft%3Druby%22+OR+%22vim%3A+set+filetype%3Druby%22&type=Code&ref=searchresults) for files that people have set as `Ruby`

**Comment by torpedro** [comment]
This seems like a really interesting project. At the top of my head I'd consider evaluating your different strategies and see how accurate each one performs on their own (you might have already done this). From there on we should have a starting point for coefficients to create a **unkown** tag, when the strategies return disjoint results.
I'd also consider returning confidence scores for all strategies if possible. For example I could imagine for the filename strategy to return for ".for" files [60% FORTRAN, 20% Forth, 20% Text]. These values could be learned form a similar dataset with which you train your classifier. What do you think?
Have you at some point considered not only relying on the file itself but also the files within the same project? I'd assume that there would be some information to gain there. Simple example: If a project contains .cpp files then I'd assume that all .h would be C++ rather than C. This is a rather trivial example, but I can imagine there might be correlations between the language of surrounding files and the file we're analyzing now. Maybe you already have some experience?

Thanks! Really interesting project!

**Comment by Sayan-Paul** [comment]
If we want to come up with a weighted scoring for all the strategies we first need to analyse the relevance of each strategy as generalized language identifier. For example, Shebangs will not be a efficient strategy for non-Unix platform based projects and also not all languages use it. Similar case is for modelines too. So like this we need analyse how individual system/strategy perform in a general language model perspective. Secondly, as mentioned we can try to implement a more sophisticated learning/classifier system to get better results.

**Comment by jwerlley** [comment]
.. cool .. We could use recombination algorithms such as genetic algorithm (GA) to test the various techniques Statistics have greater Highscore hit. ¬.¬

**Comment by ScrimpyCat** [comment]
I'm not a participant in GSoC, but I have some ideas/suggestions. 

One of the main issues I see with the classifier is how it currently gives weights to tokens/how it works with the tokenizer. Things that end up in the `samples.json` that aren't actually useful for certain languages (variable names, etc.), but instead have the negative effect of possibly bumping up code that is being classified to lean in its favour.

A lot of this was happening back with https://github.com/github/linguist/issues/1649 which was fixed through a heuristic in https://github.com/github/linguist/pull/1652 And this seems like the way it would have to be unless the tokenizer is changed. In https://github.com/github/linguist/pull/1027 the tokenizer was changed to try include more common number literal formats (avoiding random `e`, `+`, `-`, `f`, etc. from being left behind and being used as a token and given a weight), but there's still more that would need to be done (as mentioned with variable names). All in all I'm not sure if generating weights from code samples is really the right way to go. It's good for being little effort to add support for a new language, but as far as perfecting its language detection, that seems much trickier to do.

The heuristics on the other hand are a much better approach as we can be specific about what is and isn't valid, the downside of it is that it takes more work, people need to know ruby to contribute, may end up having a negative impact on performance if it grows a lot.

So I was thinking what if another stage was added, similar to heuristics but a simpler/stricter format to add detection of things that are valid or invalid. e.g. This could be a combination of rules/truthes about a specific language that could be added. Such as adding a rule for a string specifying what syntax makes up a comment (let's say anything between /* */), and then specifying that any character is valid inside that comment; while another rule might say only a subset of characters are valid, etc. And so rules could then just be added as needed overtime.

**Comment by pchaigno** [comment]
> Things that end up in the samples.json that aren't actually useful for certain languages (variable names, etc.), but instead have the negative effect of possibly bumping up code that is being classified to lean in its favour.

This issue can be solved by adding more samples, it would reduce the weight of variable names for instance. #2117 is a first step toward this.

> All in all I'm not sure if generating weights from code samples is really the right way to go. It's good for being little effort to add support for a new language, but as far as perfecting its language detection, that seems much trickier to do.

It's also much better for performance!

The idea you propose is, imo, close to one of the ideas suggested for GSoC, *[Use grammar bundles to tokenize languages](https://github.com/github/gsoc#use-grammar-bundles-to-tokenize-languages)*. I'm curious how it would impact performance though.

**Comment by arfon** [comment]
Closing this as stale (it's still a good idea though :smile:)
